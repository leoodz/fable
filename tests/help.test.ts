import { assertEquals } from '$std/assert/mod.ts';

import { assertSnapshot } from '$std/testing/snapshot.ts';

import help from '../src/help.ts';

Deno.test('/help', async (test) => {
  await test.step('navigation', () => {
    const message = help.pages({ userId: 'user_id', index: 0 });

    assertEquals(message.json().data.components[0].components[0], {
      custom_id: 'help==9=prev',
      label: 'Prev',
      style: 2,
      type: 2,
    });

    assertEquals(message.json().data.components[0].components[1], {
      custom_id: '_',
      disabled: true,
      label: '1/10',
      style: 2,
      type: 2,
    });

    assertEquals(message.json().data.components[0].components[2], {
      custom_id: 'help==1=next',
      label: 'Next',
      style: 2,
      type: 2,
    });
  });

  await test.step('pages', async () => {
    for (let i = 0; i < 10; i++) {
      const message = help.pages({ userId: 'user_id', index: i });

      await assertSnapshot(test, message.json());
    }
  });
});
