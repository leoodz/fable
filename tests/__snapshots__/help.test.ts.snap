export const snapshot = {};

snapshot[`/help 1`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==9=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "1/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==1=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            custom_id: "gacha=user_id",
            label: "/gacha",
            style: 2,
            type: 2,
          },
          {
            custom_id: "now=user_id",
            label: "/now",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "1.",
        },
        description: "Before anything: You will need characters.

__You get a free pull every 30 minutes (maxed at 5 pulls)__. Use \`/now\` to check how many pulls you have and how much time is left until your next free pull.

__Characters are exclusive__, once a player on this server finds a character, no one else can find that character in gacha again.

__Characters vary in ranks from 1<:smolstar:1107503653956374638>to 5<:smolstar:1107503653956374638>__ A character rating is based on their popularity (a character rating cannot be raised).

__Gacha rates depend on multiple variables__, but generally speaking the chances of a 5<:smolstar:1107503653956374638>appearing is equal or less than 1%.

> *\`/q\` allows you to do gacha pulls with no animations*
​",
        footer: {
          text: "aliases: /w, /q",
        },
        title: "\`/gacha\`",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 2`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==0=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "2/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==2=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "2.",
        },
        description: "Sacrifice characters to pull a new character with a higher rating.

__Your party members and likes will never be sacrificed.__ To protect characters from being merged, add them to your likeslist with \`/like\`

__It takes 5 characters with a lower rating__ to pull a new character

__Fable will always auto-merge your lowest-rated characters first.__

Example: \`/merge target: 3\` will sacrifice __25 of your 1<:smolstar:1107503653956374638>characters__
or __5 of your 2<:smolstar:1107503653956374638>characters.__

> *Sacrificed characters will return to being available in \`/gacha\`*
​",
        footer: {
          text: "aliases: /synthesis",
        },
        title: "\`/merge\`",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 3`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==1=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "3/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==3=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            custom_id: "cstars=5=user_id",
            label: "/collection stars 5",
            style: 2,
            type: 2,
          },
          {
            custom_id: "cstars=4=user_id",
            label: "/collection stars 4",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "3.",
        },
        description: "You will need a party to use combat features

__You can't have more than 1 party__ or save your current party to a preset and switch between presets.

__The more your party members are familiar with each other the better they preform__. This is called synergy, five party members from the same anime are better than 5 members from a mix of different media.

\`/party view\` to see your current party.

\`/party assign name: spot:\` to assign a character you have to your party,
leaving the \`spot:\` empty will assign the character to the first empty spot or override the last spot.

> *\`/collection\` can be used to browse your characters and select your potential party members*
​",
        footer: {
          text: "aliases: /team, /p",
        },
        title: "\`/party\`",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 4`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==2=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "4/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==4=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Stealing",
        },
        description: "When all negotiations fails, it's fine to take what you want by force, right?

__You can't steal characters assigned as party members__, meaning you can protect your most precious characters by adding them to your party.

__Inactive users are easier to steal from.__ Stay active by using \`/gacha\` or \`/q\` once a day.

__Inactive users lose their party protection__ after **4** days.

__The higher the character's rating the harder it is to steal.__ 5<:smolstar:1107503653956374638>characters have a base success rate of 1%.

__Steal has a cooldown of **3** days__ regardless of the outcome of the attempt.

> Staying active is the best defense against stealing.
​",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 5`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==3=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "5/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==5=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Combat",
        },
        description: "Prepare your characters for combat using \`/stats\`

__Every character gets a starting amount of unclaimed stat points__ based on their star rating.

__You can distribute those unclaimed points in 3 different base stats:__ strength, stamina, and agility.

__Strength is the damage__ the character does when they land a hit.

__Agility is for dodging__ attacks if your agility is higher than your enemy they can't hit you.

__Stamina is depleted__ every time your character attacks or dodges, if a character fully runs out of stamina they can no longer attack and all attack on them hit automatically.

> Finding a good balance between all 3 stats is the most recommended route.

__**NOTE: Combat is currently experimental and subject to change**__
​",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 6`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==4=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "6/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==6=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            custom_id: "now=user_id",
            label: "/vote",
            style: 2,
            type: 2,
          },
          {
            custom_id: "buy=bguaranteed=user_id=5",
            label: "/buy guaranteed 5",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Shop",
        },
        description: "The shop is a way to purchase extra pulls and guaranteed pulls.

__You get 1 daily token every day__, automatically added after your first \`/gacha\` of the day.

__Voting for Fable on Top.gg awards 1 token__, you can vote once every 12 hours.

__Voting on weekends awards 2 tokens__ (_Fridays, Saturdays and Sundays_).

\`/buy normal\` gets you extra pulls, they are the exact same as the ones you get every 30 minutes.

\`/buy guaranteed\` guarantees you get a character from one specific star rating.

Example: \`/buy guaranteed 5\` will guarantee you get one random 5<:smolstar:1107503653956374638>character.

3<:smolstar:1107503653956374638>cost **4** tokens
4<:smolstar:1107503653956374638>cost **12** tokens
5<:smolstar:1107503653956374638>cost **28** tokens

> Use the guaranteed pulls you bought by calling \`/pull stars:\`
​",
        footer: {
          text: "aliases: /shop",
        },
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 7`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==5=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "7/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==7=next",
            label: "Next",
            style: 2,
            type: 2,
          },
          {
            label: "GitHub",
            style: 5,
            type: 2,
            url: "https://github.com/ker0olos/fable",
          },
          {
            label: "Discord Support Server",
            style: 5,
            type: 2,
            url: "https://discord.gg/H69RVBxeYY",
          },
        ],
        type: 1,
      },
      {
        components: [
          {
            label: "Donate",
            style: 5,
            type: 2,
            url: "https://ko-fi.com/ker0olos",
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Roadmap",
        },
        description: "__**Releasing in the near future (~3 months)**__

**[Battles](https://github.com/ker0olos/fable/issues/74)**

**[Leveling Up Characters](https://github.com/ker0olos/fable/issues/64)**

**[Player vs. Environment](https://github.com/ker0olos/fable/issues/58)**
",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 8`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==6=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "8/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==8=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Essential Commands",
        },
        description: "- \`/gacha\` \`/w\`: _start a new gacha pull_
- \`/q\`: _start a gacha pull but with no animations_
- \`/pull\` \`/guaranteed\`: _use a guaranteed pull you have_

- \`/now\` \`/vote\` \`/tu\`: _check what you can do right now_
- \`/search\` \`/anime\` \`/manga\` \`/series\`: _search for a specific series_
- \`/character\` \`/char\`: _search for a specific character_

- \`/merge\` \`/synthesize\`: _merge characters together to pull a new character_

- \`/party view\` \`/team view\` \`/p view\`: _view your current party_
- \`/party assign\` \`/team assign\` \`/p assign\`: _assign a character to your party_
- \`/party swap\` \`/team swap\` \`/p swap\`: _swap characters spots with each others_
- \`/party remove\` \`/team remove\` \`/p remove\`: _remove a character from your party_

- \`/collection stars\` \`/coll stars\` \`/mm stars\`: _view user stars_
- \`/collection media\` \`/coll media\` \`/mm media\`: _view user characters in a specific series_
- \`/collection sum\` \`/coll sum\` \`/mm sum\`: _view the sum numbers of a user collection_

- \`/steal\`: _steal a character from another user_
- \`/trade\` \`/offer\`: _trade characters with another user_
- \`/give\` \`/gift\`: _give characters to another user_",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 9`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==7=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "9/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==9=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Other Commands",
        },
        description: "- \`/nick\`: _change the nickname of a character_
- \`/image\` \`/custom\`: _change the image of a character_

- \`/found\` \`/owned\`: _list all characters found from a specific series_

- \`/likeslist\`: _list all user likes_

- \`/like\` \`/protect\` \`/wish\`: _like a character to be notified if someone finds them_
- \`/unlike\`: _remove character from your likes_
- \`/likeall\`: _like a series to be notified if someone finds any character from it_
- \`/unlikeall\`: _remove a series from your likes_

- \`/buy normal\` \`/shop normal\`: _use tokens to buy normal pulls_
- \`/buy guaranteed\` \`/shop guaranteed\`: _use tokens to buy pulls with a specific guaranteed rating_

- \`/automerge min\`: _merge your characters together to pull the lowest rating possible_
- \`/automerge max\`: _merge your characters together to pull the highest rating possible_

- \`/packs\`: _list all the packs on a server_

- \`/logs\`: _list user last few found characters_",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;

snapshot[`/help 10`] = `
{
  data: {
    attachments: [],
    components: [
      {
        components: [
          {
            custom_id: "help==8=prev",
            label: "Prev",
            style: 2,
            type: 2,
          },
          {
            custom_id: "_",
            disabled: true,
            label: "10/10",
            style: 2,
            type: 2,
          },
          {
            custom_id: "help==0=next",
            label: "Next",
            style: 2,
            type: 2,
          },
        ],
        type: 1,
      },
    ],
    embeds: [
      {
        author: {
          name: "Admin Commands",
        },
        description: "_> \`/community\` requires \`Manage Server\` permission by default, but that can be changed in the \`Integrations\` section of your server settings, although we never recommend that._
- \`/community popular\`: _browse the most popular community packs_
- \`/community install\`: _install a community pack_
- \`/community uninstall\`: _uninstall a community pack_

> To create your own community pack visit <https://packs.deno.dev>",
        type: "rich",
      },
    ],
  },
  type: 4,
}
`;
